File: dev/scripts/taskgraph.py
#!/usr/bin/env python3
import argparse
import json
import sys
import uuid
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
GRAPH_DIR = ROOT / ".a2r" / "graphs"
WIH_DIR = ROOT / ".a2r" / "wih"
RUN_STATE_DIR = ROOT / ".a2r" / "run_state"
RECEIPTS_DIR = ROOT / ".a2r" / "receipts"
ARTIFACTS_DIR = ROOT / ".a2r" / "artifacts"

BEADS_KEYS = [
    "wih_version",
    "task_id",
    "graph_id",
    "title",
    "blocked_by",
    "write_scope",
    "tools",
    "acceptance",
]

AGENT_PROFILES_PATH = ROOT / "agent" / "agent_profiles.json"

NODE_STATUS_PENDING = "PENDING"
NODE_STATUS_BLOCKED = "BLOCKED"
NODE_STATUS_RUNNING = "RUNNING"
NODE_STATUS_SUCCEEDED = "SUCCEEDED"
NODE_STATUS_FAILED = "FAILED"
NODE_STATUS_SKIPPED = "SKIPPED"


def fail(message: str) -> None:
    print(f"ERROR: {message}", file=sys.stderr)
    sys.exit(1)


def load_json(path: Path) -> dict:
    try:
        return json.loads(path.read_text())
    except Exception as exc:
        fail(f"Failed to read {path}: {exc}")


def load_graph(graph_id: str) -> dict:
    graph_path = GRAPH_DIR / f"{graph_id}.json"
    if not graph_path.exists():
        fail(f"Graph not found: {graph_path}")
    graph = load_json(graph_path)
    if graph.get("graph_id") != graph_id:
        fail(f"Graph id mismatch in {graph_path}")
    return graph


def load_wih(wih_path: str) -> dict:
    rel_path = wih_path.lstrip("/")
    path = ROOT / rel_path
    if not path.exists():
        fail(f"WIH not found: {path}")
    return load_json(path)


def load_agent_profiles() -> set[str]:
    profiles = load_json(AGENT_PROFILES_PATH)
    items = profiles.get("profiles", [])
    if not items:
        fail("Agent profiles list is empty")
    ids = set()
    for profile in items:
        pid = profile.get("id")
        if not pid:
            fail("Agent profile missing id")
        ids.add(pid)
    return ids


def require_agent_profile(agent_profile_id: str) -> None:
    profiles = load_agent_profiles()
    if agent_profile_id not in profiles:
        fail(f"Unknown agent profile: {agent_profile_id}")


def validate_wih_node(node: dict, wih: dict, graph_id: str) -> None:
    if wih.get("task_id") != node.get("task_id"):
        fail(f"WIH task_id mismatch for {node.get('task_id')}")
    if wih.get("graph_id") != graph_id:
        fail(f"WIH graph_id mismatch for {node.get('task_id')}")

    graph_blocked = node.get("blocked_by", [])
    wih_blocked = wih.get("blocked_by", [])
    if graph_blocked != wih_blocked:
        fail(f"blocked_by mismatch for {node.get('task_id')}")

    beads = wih.get("beads")
    if not isinstance(beads, dict):
        fail(f"Beads envelope missing for {node.get('task_id')}")

    for key in BEADS_KEYS:
        if beads.get(key) != wih.get(key):
            fail(f"Beads.{key} mismatch for {node.get('task_id')}")


def install_run(graph_id: str, agent_profile_id: str) -> dict:
    require_agent_profile(agent_profile_id)
    graph = load_graph(graph_id)
    nodes = graph.get("nodes", [])
    if not nodes:
        fail(f"Graph has no nodes: {graph_id}")

    for node in nodes:
        wih = load_wih(node.get("wih_path", ""))
        validate_wih_node(node, wih, graph_id)

    run_id = str(uuid.uuid4())
    now = datetime.now(timezone.utc).isoformat()

    node_states = []
    runnable = []
    for node in nodes:
        blocked = node.get("blocked_by", [])
        status = NODE_STATUS_BLOCKED if blocked else NODE_STATUS_PENDING
        if status == NODE_STATUS_PENDING:
            runnable.append(node.get("task_id"))
        node_states.append({
            "task_id": node.get("task_id"),
            "status": status,
            "last_receipt_id": ""
        })

    run_state = {
        "run_id": run_id,
        "graph_id": graph_id,
        "status": "RUNNING",
        "started_at": now,
        "node_states": node_states,
        "resume_cursor": {
            "last_node": "",
            "next_nodes": sorted(runnable)
        },
        "receipts": [],
        "artifacts": []
    }

    RUN_STATE_DIR.mkdir(parents=True, exist_ok=True)
    run_path = RUN_STATE_DIR / f"{run_id}.json"
    run_path.write_text(json.dumps(run_state, indent=2))
    write_agent_execution_receipt(
        run_id,
        graph_id,
        agent_profile_id,
        "INSTALLED",
        run_path,
    )

    return {
        "run_id": run_id,
        "graph_id": graph_id,
        "run_state_path": str(run_path)
    }


def resume_run(run_id: str, agent_profile_id: str) -> dict:
    require_agent_profile(agent_profile_id)
    run_path = RUN_STATE_DIR / f"{run_id}.json"
    if not run_path.exists():
        fail(f"Run state not found: {run_path}")
    run_state = load_json(run_path)

    graph_id = run_state.get("graph_id")
    if not graph_id:
        fail(f"Run state missing graph_id: {run_path}")

    graph = load_graph(graph_id)
    nodes = graph.get("nodes", [])
    if not nodes:
        fail(f"Graph has no nodes: {graph_id}")

    node_states = run_state.get("node_states", [])
    state_by_task = {state.get("task_id"): state for state in node_states}
    for node in nodes:
        task_id = node.get("task_id")
        if task_id not in state_by_task:
            fail(f"Run state missing node state for {task_id}")

        wih = load_wih(node.get("wih_path", ""))
        validate_wih_node(node, wih, graph_id)

    receipt_dir = RECEIPTS_DIR / run_id
    receipt_nodes = set()
    receipt_ids = []
    if receipt_dir.exists():
        for receipt_path in receipt_dir.glob("*.json"):
            receipt = load_json(receipt_path)
            receipt_run_id = receipt.get("run_id")
            if receipt_run_id and receipt_run_id != run_id:
                fail(f"Receipt run_id mismatch in {receipt_path}")
            node_id = receipt.get("node_id") or receipt.get("task_id")
            if node_id:
                receipt_nodes.add(node_id)
            receipt_id = receipt.get("receipt_id") or receipt_path.stem
            receipt_ids.append(receipt_id)

    for state in node_states:
        if state.get("status") == NODE_STATUS_SUCCEEDED:
            if state.get("task_id") not in receipt_nodes:
                fail(f"Missing receipt for completed node {state.get('task_id')}")

    runnable = []
    for node in nodes:
        task_id = node.get("task_id")
        state = state_by_task.get(task_id, {})
        status = state.get("status")
        if status in [NODE_STATUS_SUCCEEDED, NODE_STATUS_FAILED, NODE_STATUS_SKIPPED]:
            continue

        deps = node.get("blocked_by", [])
        deps_missing = [dep for dep in deps if dep not in receipt_nodes]
        if deps_missing:
            if status == NODE_STATUS_RUNNING:
                fail(f"Node {task_id} running without dependency receipts")
            state["status"] = NODE_STATUS_BLOCKED
            continue

        if status == NODE_STATUS_BLOCKED:
            state["status"] = NODE_STATUS_PENDING
        if state["status"] == NODE_STATUS_PENDING:
            runnable.append(task_id)

    run_state["receipts"] = sorted(receipt_ids)
    run_state["resume_cursor"] = {
        "last_node": run_state.get("resume_cursor", {}).get("last_node", ""),
        "next_nodes": sorted(runnable)
    }

    run_path.write_text(json.dumps(run_state, indent=2))
    write_agent_execution_receipt(
        run_id,
        graph_id,
        agent_profile_id,
        "RESUMED",
        run_path,
    )

    return {
        "run_id": run_id,
        "graph_id": graph_id,
        "next_nodes": sorted(runnable)
    }


def write_agent_execution_receipt(
    run_id: str,
    graph_id: str,
    agent_profile_id: str,
    status: str,
    run_path: Path,
) -> None:
    receipt_id = f"agent-exec-{uuid.uuid4()}"
    created_at = datetime.now(timezone.utc).isoformat()
    receipts_dir = RECEIPTS_DIR / run_id
    receipts_dir.mkdir(parents=True, exist_ok=True)
    receipt = {
        "receipt_id": receipt_id,
        "created_at": created_at,
        "run_id": run_id,
        "graph_id": graph_id,
        "agent_profile_id": agent_profile_id,
        "status": status,
        "run_state_path": f"/{run_path.as_posix()}",
        "receipts_dir": f"/{receipts_dir.as_posix()}",
        "artifacts_dir": f"/{(ARTIFACTS_DIR / run_id).as_posix()}",
        "node_receipts": [],
        "tool_receipts": [],
        "subprocess_receipts": [],
    }
    (receipts_dir / f"{receipt_id}.json").write_text(json.dumps(receipt, indent=2))


def main() -> None:
    parser = argparse.ArgumentParser(description="A2R task graph runner")
    sub = parser.add_subparsers(dest="command", required=True)

    install_parser = sub.add_parser("install", help="Create run_state for a graph")
    install_parser.add_argument("graph_id")
    install_parser.add_argument("--agent-profile", default="kernel-default")

    resume_parser = sub.add_parser("resume", help="Resume a run")
    resume_parser.add_argument("run_id")
    resume_parser.add_argument("--agent-profile", default="kernel-default")

    args = parser.parse_args()

    if args.command == "install":
        result = install_run(args.graph_id, args.agent_profile)
        print(json.dumps(result, indent=2))
        return

    if args.command == "resume":
        result = resume_run(args.run_id, args.agent_profile)
        print(json.dumps(result, indent=2))
        return

    fail("Unknown command")


if __name__ == "__main__":
    main()

File: 4-services/orchestration/kernel-service/src/taskgraph.rs
use chrono::Utc;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};
use uuid::Uuid;

const GRAPH_DIR: &str = ".a2r/graphs";
const RUN_STATE_DIR: &str = ".a2r/run_state";
const RECEIPTS_DIR: &str = ".a2r/receipts";
const ARTIFACTS_DIR: &str = ".a2r/artifacts";
const AGENT_PROFILES_PATH: &str = "agent/agent_profiles.json";

const BEADS_KEYS: [&str; 8] = [
    "wih_version",
    "task_id",
    "graph_id",
    "title",
    "blocked_by",
    "write_scope",
    "tools",
    "acceptance",
];

const STATUS_PENDING: &str = "PENDING";
const STATUS_BLOCKED: &str = "BLOCKED";
const STATUS_RUNNING: &str = "RUNNING";
const STATUS_SUCCEEDED: &str = "SUCCEEDED";
const STATUS_FAILED: &str = "FAILED";
const STATUS_SKIPPED: &str = "SKIPPED";

#[derive(Debug, Deserialize)]
struct Graph {
    graph_id: String,
    nodes: Vec<GraphNode>,
}

#[derive(Debug, Deserialize)]
struct GraphNode {
    task_id: String,
    #[serde(default)]
    blocked_by: Vec<String>,
    wih_path: String,
}

fn load_json(path: &Path) -> Result<Value, String> {
    let data = fs::read_to_string(path)
        .map_err(|e| format!("Failed to read {}: {}", path.display(), e))?;
    serde_json::from_str(&data).map_err(|e| format!("Failed to parse {}: {}", path.display(), e))
}

fn load_graph(graph_id: &str) -> Result<Graph, String> {
    let path = Path::new(GRAPH_DIR).join(format!("{}.json", graph_id));
    let data = fs::read_to_string(&path)
        .map_err(|e| format!("Graph not found: {} ({})", path.display(), e))?;
    let graph: Graph = serde_json::from_str(&data)
        .map_err(|e| format!("Failed to parse graph {}: {}", path.display(), e))?;
    if graph.graph_id != graph_id {
        return Err(format!("Graph id mismatch in {}", path.display()));
    }
    Ok(graph)
}

fn load_wih(wih_path: &str) -> Result<Value, String> {
    let rel = wih_path.trim_start_matches('/');
    let path = Path::new(rel);
    if !path.exists() {
        return Err(format!("WIH not found: {}", path.display()));
    }
    load_json(path)
}

fn value_as_str(value: &Value, key: &str) -> Result<String, String> {
    value
        .get(key)
        .and_then(|v| v.as_str())
        .map(|v| v.to_string())
        .ok_or_else(|| format!("Missing or invalid {}", key))
}

fn value_as_array(value: &Value, key: &str) -> Result<Vec<String>, String> {
    value
        .get(key)
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|item| item.as_str().map(|s| s.to_string()))
                .collect::<Vec<String>>()
        })
        .ok_or_else(|| format!("Missing or invalid {}", key))
}

fn validate_wih_node(node: &GraphNode, wih: &Value, graph_id: &str) -> Result<(), String> {
    let task_id = value_as_str(wih, "task_id")?;
    if task_id != node.task_id {
        return Err(format!("WIH task_id mismatch for {}", node.task_id));
    }
    let wih_graph_id = value_as_str(wih, "graph_id")?;
    if wih_graph_id != graph_id {
        return Err(format!("WIH graph_id mismatch for {}", node.task_id));
    }

    let wih_blocked = value_as_array(wih, "blocked_by")?;
    if wih_blocked != node.blocked_by {
        return Err(format!("blocked_by mismatch for {}", node.task_id));
    }

    let beads = wih
        .get("beads")
        .and_then(|v| v.as_object())
        .ok_or_else(|| format!("Beads envelope missing for {}", node.task_id))?;

    for key in BEADS_KEYS {
        let wih_val = wih.get(key);
        let beads_val = beads.get(key);
        if wih_val != beads_val {
            return Err(format!("Beads.{} mismatch for {}", key, node.task_id));
        }
    }

    Ok(())
}

fn load_agent_profiles() -> Result<HashSet<String>, String> {
    let data = fs::read_to_string(AGENT_PROFILES_PATH)
        .map_err(|e| format!("Agent profiles not found: {} ({})", AGENT_PROFILES_PATH, e))?;
    let json: Value = serde_json::from_str(&data)
        .map_err(|e| format!("Failed to parse agent profiles: {}", e))?;
    let profiles = json
        .get("profiles")
        .and_then(|v| v.as_array())
        .ok_or_else(|| "Agent profiles missing profiles array".to_string())?;
    let mut ids = HashSet::new();
    for profile in profiles {
        let id = profile
            .get("id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| "Agent profile missing id".to_string())?;
        ids.insert(id.to_string());
    }
    if ids.is_empty() {
        return Err("Agent profiles list is empty".to_string());
    }
    Ok(ids)
}

fn require_agent_profile(agent_profile_id: &str) -> Result<(), String> {
    let profiles = load_agent_profiles()?;
    if !profiles.contains(agent_profile_id) {
        return Err(format!("Unknown agent profile: {}", agent_profile_id));
    }
    Ok(())
}

pub fn install_run(graph_id: &str) -> Result<Value, String> {
    install_run_with_profile(graph_id, "kernel-default")
}

pub fn install_run_with_profile(graph_id: &str, agent_profile_id: &str) -> Result<Value, String> {
    require_agent_profile(agent_profile_id)?;
    let graph = load_graph(graph_id)?;
    if graph.nodes.is_empty() {
        return Err(format!("Graph has no nodes: {}", graph_id));
    }

    for node in &graph.nodes {
        let wih = load_wih(&node.wih_path)?;
        validate_wih_node(node, &wih, graph_id)?;
    }

    let run_id = Uuid::new_v4().to_string();
    let now = Utc::now().to_rfc3339();

    let mut node_states = Vec::new();
    let mut runnable = Vec::new();

    for node in &graph.nodes {
        let status = if node.blocked_by.is_empty() {
            runnable.push(node.task_id.clone());
            STATUS_PENDING
        } else {
            STATUS_BLOCKED
        };

        node_states.push(json!({
            "task_id": node.task_id,
            "status": status,
            "last_receipt_id": ""
        }));
    }

    let mut runnable_sorted = runnable.clone();
    runnable_sorted.sort();

    let run_state = json!({
        "run_id": run_id,
        "graph_id": graph_id,
        "status": "RUNNING",
        "started_at": now,
        "node_states": node_states,
        "resume_cursor": {
            "last_node": "",
            "next_nodes": runnable_sorted
        },
        "receipts": [],
        "artifacts": []
    });

    fs::create_dir_all(RUN_STATE_DIR)
        .map_err(|e| format!("Failed to create run_state dir: {}", e))?;
    let run_path = Path::new(RUN_STATE_DIR).join(format!("{}.json", run_id));
    fs::write(
        &run_path,
        serde_json::to_string_pretty(&run_state).map_err(|e| e.to_string())?,
    )
    .map_err(|e| format!("Failed to write run_state: {}", e))?;

    write_agent_execution_receipt(&run_id, graph_id, agent_profile_id, "INSTALLED", &run_path)?;

    Ok(json!({
        "run_id": run_id,
        "graph_id": graph_id,
        "run_state_path": run_path
    }))
}

pub fn resume_run(run_id: &str) -> Result<Value, String> {
    resume_run_with_profile(run_id, "kernel-default")
}

pub fn resume_run_with_profile(run_id: &str, agent_profile_id: &str) -> Result<Value, String> {
    require_agent_profile(agent_profile_id)?;
    let run_path = Path::new(RUN_STATE_DIR).join(format!("{}.json", run_id));
    if !run_path.exists() {
        return Err(format!("Run state not found: {}", run_path.display()));
    }

    let mut run_state = load_json(&run_path)?;
    let graph_id = value_as_str(&run_state, "graph_id")?;
    let graph = load_graph(&graph_id)?;

    let node_states_val = run_state
        .get("node_states")
        .and_then(|v| v.as_array())
        .ok_or_else(|| format!("Run state missing node_states: {}", run_path.display()))?;

    let mut state_by_task: HashMap<String, Value> = HashMap::new();
    for state in node_states_val {
        if let Some(task_id) = state.get("task_id").and_then(|v| v.as_str()) {
            state_by_task.insert(task_id.to_string(), state.clone());
        }
    }

    for node in &graph.nodes {
        if !state_by_task.contains_key(&node.task_id) {
            return Err(format!("Run state missing node state for {}", node.task_id));
        }
        let wih = load_wih(&node.wih_path)?;
        validate_wih_node(node, &wih, &graph_id)?;
    }

    let receipt_dir = Path::new(RECEIPTS_DIR).join(run_id);
    let mut receipt_nodes: HashSet<String> = HashSet::new();
    let mut receipt_ids: Vec<String> = Vec::new();
    if receipt_dir.exists() {
        for entry in
            fs::read_dir(&receipt_dir).map_err(|e| format!("Failed to read receipts: {}", e))?
        {
            let entry = entry.map_err(|e| format!("Failed to read receipt entry: {}", e))?;
            if !entry
                .path()
                .extension()
                .map(|ext| ext == "json")
                .unwrap_or(false)
            {
                continue;
            }
            let receipt = load_json(&entry.path())?;
            if let Some(receipt_run_id) = receipt.get("run_id").and_then(|v| v.as_str()) {
                if receipt_run_id != run_id {
                    return Err(format!(
                        "Receipt run_id mismatch in {}",
                        entry.path().display()
                    ));
                }
            }
            if let Some(node_id) = receipt.get("node_id").and_then(|v| v.as_str()) {
                receipt_nodes.insert(node_id.to_string());
            } else if let Some(task_id) = receipt.get("task_id").and_then(|v| v.as_str()) {
                receipt_nodes.insert(task_id.to_string());
            }
            let receipt_id = receipt
                .get("receipt_id")
                .and_then(|v| v.as_str())
                .map(|v| v.to_string())
                .unwrap_or_else(|| {
                    entry
                        .path()
                        .file_stem()
                        .unwrap_or_default()
                        .to_string_lossy()
                        .to_string()
                });
            receipt_ids.push(receipt_id);
        }
    }

    for state in node_states_val {
        let status = state.get("status").and_then(|v| v.as_str()).unwrap_or("");
        if status == STATUS_SUCCEEDED {
            let task_id = value_as_str(state, "task_id")?;
            if !receipt_nodes.contains(&task_id) {
                return Err(format!("Missing receipt for completed node {}", task_id));
            }
        }
    }

    let mut runnable = Vec::new();
    for node in &graph.nodes {
        let mut state = state_by_task
            .get(&node.task_id)
            .cloned()
            .unwrap_or_else(|| json!({}));
        let status = state
            .get("status")
            .and_then(|v| v.as_str())
            .unwrap_or(STATUS_PENDING);
        if [STATUS_SUCCEEDED, STATUS_FAILED, STATUS_SKIPPED].contains(&status) {
            continue;
        }

        let deps_missing: Vec<String> = node
            .blocked_by
            .iter()
            .filter(|dep| !receipt_nodes.contains(*dep))
            .cloned()
            .collect();

        if !deps_missing.is_empty() {
            if status == STATUS_RUNNING {
                return Err(format!(
                    "Node {} running without dependency receipts",
                    node.task_id
                ));
            }
            state["status"] = Value::String(STATUS_BLOCKED.to_string());
        } else {
            if status == STATUS_BLOCKED {
                state["status"] = Value::String(STATUS_PENDING.to_string());
            }
            if state["status"] == Value::String(STATUS_PENDING.to_string()) {
                runnable.push(node.task_id.clone());
            }
        }

        run_state
            .get_mut("node_states")
            .and_then(|v| v.as_array_mut())
            .ok_or_else(|| "Run state node_states missing".to_string())?
            .iter_mut()
            .for_each(|item| {
                if item.get("task_id") == Some(&Value::String(node.task_id.clone())) {
                    *item = state.clone();
                }
            });
    }

    run_state["receipts"] = Value::Array(receipt_ids.into_iter().map(Value::String).collect());
    let mut resume_cursor = run_state
        .get("resume_cursor")
        .cloned()
        .unwrap_or_else(|| json!({}));
    let mut runnable_sorted = runnable.clone();
    runnable_sorted.sort();
    resume_cursor["next_nodes"] =
        Value::Array(runnable_sorted.iter().cloned().map(Value::String).collect());
    run_state["resume_cursor"] = resume_cursor;

    fs::write(
        &run_path,
        serde_json::to_string_pretty(&run_state).map_err(|e| e.to_string())?,
    )
    .map_err(|e| format!("Failed to write run_state: {}", e))?;

    write_agent_execution_receipt(run_id, &graph_id, agent_profile_id, "RESUMED", &run_path)?;

    Ok(json!({
        "run_id": run_id,
        "graph_id": graph_id,
        "next_nodes": runnable_sorted
    }))
}

fn write_agent_execution_receipt(
    run_id: &str,
    graph_id: &str,
    agent_profile_id: &str,
    status: &str,
    run_state_path: &Path,
) -> Result<String, String> {
    let receipt_id = format!("agent-exec-{}", Uuid::new_v4());
    let created_at = Utc::now().to_rfc3339();
    let receipts_dir = Path::new(RECEIPTS_DIR).join(run_id);
    fs::create_dir_all(&receipts_dir)
        .map_err(|e| format!("Failed to create receipts dir: {}", e))?;

    let receipt = json!({
        "receipt_id": receipt_id,
        "created_at": created_at,
        "run_id": run_id,
        "graph_id": graph_id,
        "agent_profile_id": agent_profile_id,
        "status": status,
        "run_state_path": format!("/{}", run_state_path.display()),
        "receipts_dir": format!("/{}/{}", RECEIPTS_DIR, run_id),
        "artifacts_dir": format!("/{}/{}", ARTIFACTS_DIR, run_id),
        "node_receipts": [],
        "tool_receipts": [],
        "subprocess_receipts": []
    });

    let receipt_path = receipts_dir.join(format!("{}.json", receipt_id));
    fs::write(
        &receipt_path,
        serde_json::to_string_pretty(&receipt).map_err(|e| e.to_string())?,
    )
    .map_err(|e| format!("Failed to write agent execution receipt: {}", e))?;
    Ok(receipt_id)
}
